enhancements/cluster-api/cluster-autoscaler-integration-with-openshift-cluster-api.md

```diff
@@ -0,0 +1,751 @@
+`machine.openshift.io` API group) would be valid target of the `.spec.scaleTargetRef`
+field. After this enhancement is implemented, users may specify either a Machine
+API MachineSet or a Cluster API MachineSet in the `.spec.scaleTargetRef` field.
+
+For example, assume a cluster has the following MachineSet resources:
+
+```yaml
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineSet
+metadata:
+  labels:
+    machine.openshift.io/cluster-api-cluster: example-cluster
+  name: example-cluster-worker-somezone-1
+  namespace: openshift-machine-api
+spec:
+  ...
+status:
+  authoritativeAPI: MachineAPI
+```
+
+```yaml
+apiVersion: machine.openshift.io/v1beta1
+kind: MachineSet
+metadata:
+  labels:
+    machine.openshift.io/cluster-api-cluster: example-cluster
+  name: example-cluster-worker-somezone-2
+  namespace: openshift-machine-api
+spec:
+  ...
+status:
+  authoritativeAPI: ClusterAPI
+```
+
+A user could then create the following MachineAutoscaler resources:
+
+```yaml
+apiVersion: "autoscaling.openshift.io/v1beta1"
+kind: "MachineAutoscaler"
+metadata:
+  name: "worker-somezone-1"
+  namespace: "openshift-machine-api"
+spec:
+  minReplicas: 1
+  maxReplicas: 12
+  scaleTargetRef:
+    apiVersion: machine.openshift.io/v1beta1
+    kind: MachineSet
+    name: example-cluster-worker-somezone-1
+```
+
+```yaml
+apiVersion: "autoscaling.openshift.io/v1beta1"
+kind: "MachineAutoscaler"
+metadata:
+  name: "worker-somezone-2"
+  namespace: "openshift-machine-api"
+spec:
+  minReplicas: 1
+  maxReplicas: 12
+  scaleTargetRef:
+    apiVersion: cluster.x-k8s.io/v1beta1
+    kind: MachineSet
+    name: example-cluster-worker-somezone-2
+```
+
+Note that the MachineAutoscaler named "worker-somezone-1" is targeting a Machine API
+MachineSet while "worker-somezone-2" is targeting a Cluster API MachineSet. The
+Cluster Autoscaler Operator will know by the `apiVersion` field whether to look
+for the resource in the `openshift-machine-api` or `openshift-cluster-api` namespace
+respectively.
+
+To support existing user experiences and workflows, the user might also have created
+the "worker-somezone-2" MachineAutoscaler using the Machine API reference as follows:
+
+```yaml
+apiVersion: "autoscaling.openshift.io/v1beta1"
+kind: "MachineAutoscaler"
+metadata:
+  name: "worker-somezone-2"
+  namespace: "openshift-machine-api"
+spec:
+  minReplicas: 1
+  maxReplicas: 12
+  scaleTargetRef:
+    apiVersion: machine.openshift.io/v1beta1
+    kind: MachineSet
+    name: example-cluster-worker-somezone-2
+```
+
+The Cluster Autoscaler Operator will update the Machine API MachineSet resource, and
+then the MachineSet sync controller will sync the change to the Cluster API MachineSet
+resource. The sync controller will use the managed fields (i.e. `.metadata.managedFields`)
+of the specified MachineSet to determine if the Cluster Autoscaler Operator made
+changes to the annotations, and then replicate those appropriately. In this manner,
```

Comment:
If we always write to the authoritative API, I don't think managedfields is required here?